Prototype for authentication flow for  web management UI on the push bouncer

mateuszmandera opened the pull request:
This works as follows:

1. Org owner clicks "Self hosted billing" in the cogwheel menu. This takes them to `/self-hosted-billing/`.
2. The server POSTs to the push bouncer `/api/v1/remotes/server/billing` with some basic data about the user etc., authenticating with its push bouncer credentials.
3. The bouncer returns a json success payload with `billing_access_url`. This is an url to `<bouncer root domain>/remotes/billing/<signed token>` - where `<signed token>` contains the info about the user and the remote server and allows accessing the billing page for the remote server. It's considered valid for `SIGNED_ACCESS_TOKEN_VALIDITY_IN_SECONDS=24hours` by the bouncer and can be given to others to access the page, aside of the original user.
4. The remote server tells the user's browser to redirect to this url.
5. At the url, some metadata about the server and user is currently displayed, as a prototype.

No automated tests, since this design is probably gonna change as we proceed. This can be tested manually in dev env by:

1. Set `PUSH_NOTIFICATION_BOUNCER_URL = "<http://localhost:9991>"`.
2. Run `manage.py register_server` to register. Now you're good to go.

Can also be made more realistic by deploying the code to a DO droplet and setting it as your push bouncer in step 1. `https://mateuszmandera.zulipdev.org/` currently runs this code and therefore can be used as the bouncer for some quick testing.

Here's a recording of the flow:
[Screencast from 2023-11-02 16-47-43.webm](https://github.com/zulip/zulip/assets/45007152/15090015-e58a-4869-871d-2c498bbebcbb)

timabbott commented on corporate/views/remote_billing_page.py:
```diff
@@ -0,0 +1,103 @@
+        "realm_id": realm_id,
+        "user_uuid": user_uuid,
+        "remote_server_uuid": str(remote_server.uuid),
+    }
```
Should we declare these as a TypedDict so that the type-checker can know these are intended to be the same type here and when we ready them out?

mateuszmandera replied:
Made this a TypedDict

timabbott commented on corporate/views/remote_billing_page.py:
```diff
@@ -0,0 +1,103 @@
+            "realm_id": realm_id,
+            "remote_server_info": remote_server_info,
+        },
+    )
```
Am I correct that what we're doing here is just returning a JSON blob to confirm things are working, but this would in the future just be rendering a template with these fields as context?

mateuszmandera replied:
Correct. Now I replaced this response with a dummy html template that displays some user and server info

mateuszmandera commented on zilencer/auth.py:
```diff
@@ -82,7 +83,8 @@ def validate_remote_server(
         raise RemoteServerDeactivatedError
 
-    if get_subdomain(request) != Realm.SUBDOMAIN_FOR_ROOT_DOMAIN:
+    if get_subdomain(request) != Realm.SUBDOMAIN_FOR_ROOT_DOMAIN and not settings.DEVELOPMENT:
```
Tweaked the commit

mateuszmandera commented on corporate/views/remote_billing_page.py:
```diff
@@ -0,0 +1,103 @@
+    if not user.is_realm_owner:
+        raise OrganizationOwnerRequiredError
+
+    post_data: Dict[str, Union[str, int]] = {"realm_id": user.realm_id, "user_uuid": str(user.uuid)}
```
Added a bunch more fields
```
    realm_info = get_realms_info_for_push_bouncer(user.realm_id)[0]

    post_data = {
        "user_uuid": orjson.dumps(str(user.uuid)).decode(),
        "user_email": orjson.dumps(user.delivery_email).decode(),
        "user_full_name": orjson.dumps(user.full_name).decode(),
        "realm_info": orjson.dumps(realm_info).decode(),
    }
```

mateuszmandera commented:
@timabbott Pushed a revised version with above feedback addressed, the `RemoteRealm` logic handling and the couple other discussed points (`is_billing_user` and sending more user info). Recording of how this looks now (well, all that's changed is that it's called "Billing" now in the menu and we show an HTML page at the end):

[Uploading Screencast from 2023-11-15 23-00-50.webm…]()



Note: this is currently stacked on top of #27677 right now, for the realm info uploading function - but i can extract the function into a separate commit  if we wanna hold off on the register_server thing for now.

Here's a more details changelog compared to what this was before, in case that's helpful:
- Addressed PR review feedback
- Renamed “Self hosted billing” in the menu to just “Billing” as discussed in the thread
- Added both `RemoteRealm` and no `RemoteRealm` handling (for pre-8.0 servers
- Added more info to be sent to the auth API by the remote server (about realm and user)
- The user info is now displayed in an html template, to make this prototype a bit closer to the target. We’ll have a proper HTML page for billing here in the future.
- Moved the `self-hosted-billing/` remote server-side endpoint out of `corporate/urls.py` since normal servers don’t run `corporate/` i think, so that was wrong. Now this endpoint is registered in `zproject/urls.py` and the function definition lives in `zerver/views/push_notifications.py`
- Added handling of the “no RemoteRealm matching the realm.uuid provided” error case, on both the bouncer (return proper error response) and the remote server (recognize it and upload its realm info to the `remotes/server/analytic` endpoint to create the registration and then re-try the auth request)
- Support changing `is_billing_admin` via `manage.py change_user_role`
- `is_billing_admin` permission on a remote server now allows the user to go through the remote billing authentication flow

So far this I tested this  manually. If this looks close enough to how we want it, I could get started on writing tests.

mateuszmandera commented on corporate/views/remote_billing_page.py:
```diff
@@ -0,0 +1,140 @@
+    ),
+) -> HttpResponse:
+    remote_realm = None
+    if "uuid" in realm_info:
+        # This codepath is for servers version 8.0+, as they submit the realm uuid,
+        # allowing us to tie this process to a RemoteRealm.
```
I also considered whether the remote server should just send its `version` to the endpoint, and we'd make the decision based on that. But I think current approach is cleaner - a pre-8.0 server won't have `realm.uuid` fields, so it won't send them, and that's how this endpoint know it's dealing with <8.0.

timabbott replied:
Yeah that seems OK. Also a pre-8.0 server won't have `realm_info` at all, probably.

mateuszmandera replied:
Ah okay, I thought we'd backport a limited version of this flow to 7.x or something, so it'd still be similar, just wouldn't send the uuid, but would send other stuff

timabbott replied:
If we're backporting to 7.x, we might just backport a migration to create `realm.uuid` too.

mateuszmandera replied:
Hmm, okay, I cut out this check for now and have the flow assume we have `RemoteRealm`. It's a simple change to bring back in some shape if needed once we know exactly what we're doing with 7.x

timabbott replied:
Sounds good.

mateuszmandera commented on zerver/lib/exceptions.py:
```diff
@@ -570,3 +571,16 @@ def __init__(self, msg: str, error_type: str) -> None:
+
+class MissingRemoteRealmError(JsonableError):
+    code: ErrorCode = ErrorCode.MISSING_REMOTE_REALM
+    http_status_code = 403
```
Does this status code or a different one (like the default 400) make more sense?

timabbott replied:
I'm not sure, I guess 404 could be most appropriate depending on how things get hooked up.

timabbott replied:
I think after looking at this a bit more, 400 is probably most appropriate.

timabbott commented:
> Note: this is currently stacked on top of #27677 right now, for the realm info uploading function - but i can extract the function into a separate commit if we wanna hold off on the register_server thing for now.

Yeah extracting that sounds good -- I'm not convinced we want the exact `register_server` UI there but I think we're pretty likely to use this function in at least one place.

timabbott commented on zerver/tests/test_push_notifications.py:
```diff
@@ -443,12 +443,13 @@ def test_register_remote_push_user_params(self) -> None:
-            dict(user_id=user_id, token_kind=token_kind, token=token),
-            subdomain="zulip",
-        )
+        with self.settings(DEVELOPMENT=False):
```
Should this still be overriding `DEVELOPMENT` rather than `DEVELOPMENT_DISABLE_PUSH_BOUNCER_DOMAIN_CHECK`? And do we want to set that in `test_settings.py`?

mateuszmandera replied:
Oops, thanks for catching - I think this commit shouldn't be modifying the test at all anymore. I wouldn't set `DEVELOPMENT_DISABLE_PUSH_BOUNCER_DOMAIN_CHECK = True` in `test_settings`, because our tests are written with the domain check being active in mind - e.g. this original test that was modifying here, which tests the check.

My current idea here was to just have this `DEVELOPMENT_DISABLE_PUSH_BOUNCER_DOMAIN_CHECK` line in `dev_settings` that we can quickly uncomment when we want to do simple bouncer testing in dev env - and to not change any current behaviors by default.

timabbott replied:
OK, that works.

timabbott commented on zerver/management/commands/change_user_role.py:
```diff
@@ -86,9 +88,17 @@ def handle(self, *args: Any, **options: Any) -> None:
+            elif not user.is_billing_admin and not options["grant"]:
+                raise CommandError("User is not a billing admin for this realm.")
+
+            do_change_is_billing_admin(user, options["grant"])
```
This part looks great, opened as https://github.com/zulip/zulip/pull/27729 so it can merge without delay and shrink this a bit.

timabbott commented on corporate/views/remote_billing_page.py:
```diff
@@ -0,0 +1,140 @@
+                ("uuid_owner_secret", check_string),
+            ],
+        )
+    ),
```
This might be cleaner with `@typed_endpoint`, since then we can declare the type of the `realm_info` once and then have both this endpoint and `remote_server_post_analytics` process it into a shared data class. 

Though I think that only makes sense if my thought in https://github.com/zulip/zulip/pull/27575/commits/b820be56926e79e01de3560917e50134a2e33e8a#r1395066450 that this code path will be for 8.0+ only is correct, and thus we don't need a different type here.

mateuszmandera replied:
Okay, I guess that necessitates changing `remote_server_post_analytics` to be a `@typed_endpoint` too, right? I pushed a prep commit to do that - and then have this new `remote_server_billing_entry` re-use the `RealmDataForAnalytics` structure in the prototype commit, as a `@typed_endpoint` as well

timabbott replied:
Yeah this looks pretty clean. I guess we can also push some of the typing down into `validate_incoming_table_data` as a follow-up, but that's independent of the rest of what we're doing.

timabbott commented:
The prep commits look great; I've extracted them to https://github.com/zulip/zulip/pull/27777 so they can merge once CI passes.

mateuszmandera commented:
Pushed a commit adding an iteration on this to make a skeleton for the scheme of:

* Step one, the current `remote_server_billing_entry`, would remain an endpoint authenticated by the remote server's credentials that returns a signed URL that `/self-hosted-billing` on your self-hosted server can serve a redirect to, as it currently works.
* The end result of that redirect -  a page which just checks the signed object and creates a session cookie logging the customer in  and then redirects to the next step (which may vary depending on the realm's state; one without anything setup goes to /plans, and one with a plan goes to /billing).
* The actual `/billing` URL or whatever for that realm has access authenticated using that session cookie

mateuszmandera commented on zilencer/urls.py:
```diff
@@ -42,5 +39,4 @@
 urlpatterns = [
     path("api/v1/", include(push_bouncer_patterns)),
     path("api/v1/", include(billing_patterns)),
-    path("remotes/billing/<signed_billing_access_token>", remote_server_billing_ui),
```
This now turns into `"remote-billing-login/<signed_billing_access_token>", remote_server_billing_finalize_login`. Moved it to `corporate/urls.py`, since that's where our user-facing billing endpoints live. 

I'm keeping `remote_server_billing_entry` in `zilencer/urls.py`, because that's essentially a bouncer endpoints - users don't interact with it - remote servers do, and their authenticate with their bouncer credentials. Hopefully this convention makes sense.

mateuszmandera commented on corporate/views/remote_billing_page.py:
```diff
@@ -76,6 +76,22 @@ def remote_server_billing_ui(
 
+    request.session["remote_billing_identity"] = identity_dict
+
+    # TODO: Figure out redirects based on whether the realm/server already has a plan
+    # and should be taken to /billing or doesn't have and should be taken to /plans.
+    # For now we're only implemented the case where we have the RemoteRealm, and we take
+    # to /plans.
+    remote_realm_uuid = identity_dict["remote_realm_uuid"]
+
+    return HttpResponseRedirect(reverse("remote_billing_plans_realm", args=(remote_realm_uuid,)))
```
Here we'd have proper logic to redirect to a server vs realm and /plans vs /billing endpoint depending on the conditions.

timabbott replied:
Sounsd great!

timabbott commented on corporate/views/remote_billing_page.py:
```diff
@@ -0,0 +1,175 @@
+
+@typed_endpoint
+def remote_billing_page_realm(request: HttpRequest, *, realm_uuid: PathOnly[str]) -> HttpResponse:
+    return remote_billing_page_common(request, realm_uuid=realm_uuid, server_uuid=None)
```
I think we probably want to add a decorator to all these endpoints enforcing that these URLs are all on `SELF_HOSTING_SUBDOMAIN`, a new default-None setting that we can create with perhaps a `selfhosting` value defined in `dev_settings.py`. We should just have them 404 if this isn't the right subdomain, since that is what would happen if this feature didn't exist and you tried to visit these URLs on a random Zulip server.

We might also need to add checks in places like `zerver/middleware.py` where we call into `SOCIAL_AUTH_SUBDOMAIN`, e.g. this block:
```
        subdomain = get_subdomain(request)                                                              
        if subdomain == settings.SOCIAL_AUTH_SUBDOMAIN:                                                 
            # Realms are not supposed to exist on SOCIAL_AUTH_SUBDOMAIN.                                
            return None                                                                                 
```

So maybe it makes sense for this bit to be a preparatory commit defining a `SELF_HOSTING_SUBDOMAIN` setting, adding checks like that, and then this file can define a little decorator to 404 if you're not on that domain that we'd use on all these methods.

mateuszmandera replied:
Added the prep commit and decorator

timabbott commented on zerver/views/push_notifications.py:
```diff
@@ -101,3 +112,39 @@ def send_test_push_notification_api(
+
+
+@zulip_login_required
+def self_hosted_billing(request: HttpRequest) -> HttpResponse:
```
Let's call function something like `self_hosting_auth_redirect`

timabbott commented on zerver/views/push_notifications.py:
```diff
@@ -101,3 +112,39 @@ def send_test_push_notification_api(
+@zulip_login_required
+def self_hosted_billing(request: HttpRequest) -> HttpResponse:
+    if not uses_notification_bouncer():
+        raise JsonableError(_("This server doesn't use the push notification bouncer."))
```
Hmm, since this is a redirect to an HTML page, we probably don't want a JSON format error here? One could argue this should just return a standard Django 404 error, but probably we can resolve this the same way we address the `TODO` just below?

(Both errors should be basically unreachable)

mateuszmandera replied:
Yeah with the change in `gear_menu: Add proper logic for when to display remote billing`, this should be inaccessible in these cases - unless the user directly enters the URL in their browser. So 404 probably makes the most sense in the "not using bouncer" case - and in the other one, perhaps a json error is better than adding a new HTML template just for this? I added a commit with a comment like this:
```
        # We may want to replace this with an html error page at some point,
        # but this endpoint shouldn't be accessible via the UI to an unauthorized
        # user - and they need to directly enter the URL in their browser. So a json
        # error may be sufficient.
        raise OrganizationOwnerRequiredError
```

timabbott replied:
Sounds good!

timabbott commented on zerver/views/push_notifications.py:
```diff
@@ -101,3 +112,39 @@ def send_test_push_notification_api(
+        "user_email": orjson.dumps(user.delivery_email).decode(),
+        "user_full_name": orjson.dumps(user.full_name).decode(),
+        "realm_info": orjson.dumps(dict(realm_info)).decode(),
+    }
```
What do you think about making this instead a `user` object with `uuid/email/full_name` fields and a `realm` object with the realm details? To me, that feels a bit cleaner/simpler and more extensible.

mateuszmandera replied:
Yes, that's a better design, made the corresponding change. So now this adds a `UserDataForRemoteBilling(BaseModel)` to have this API param properly typed

timabbott replied:
Looks great!

timabbott commented on corporate/views/remote_billing_page.py:
```diff
@@ -0,0 +1,175 @@
+    except RemoteRealm.DoesNotExist:
+        # This error will prod the remote server to submit its realm info, which
+        # should lead to the creation of this missing RemoteRealm registration.
+        raise MissingRemoteRealmError
```
Should we check the `RemoteRealm` `uuid_owner_secret` here, which is present in this API request? I think the answer might be that we intend to use that secret only for transferring a `RealmRealm` to another `RemoteZulipServer`. But it might be worth a comment.

mateuszmandera replied:
Closing this based on the czo discussion

timabbott commented on corporate/views/remote_billing_page.py:
```diff
@@ -0,0 +1,175 @@
+    except signing.BadSignature:
+        raise JsonableError(_("Invalid billing access token."))
+
+    request.session["remote_billing_identity"] = identity_dict
```
Is there anything special we need to do with Django sessions to isolate this session variable so that things don't get mixed up with other realms for someone who runs multiple Zulip servers?

One theory would be to make the structure something like `request.session["self_hosting_realms"][remote_realm_uuid] = identity_dict`, so that we'd only look at parts of this session when the URL contains that realm UUID.

We also will need to consider the revocation/expiry story for these sessions.

This detail probably deserves a dedicated chat.zulip.org thread.

mateuszmandera replied:
> One theory would be to make the structure something like request.session["self_hosting_realms"][remote_realm_uuid] = identity_dict, so that we'd only look at parts of this session when the URL contains that realm UUID.

This sounds like a good idea to me - may help us properly support being logged into multiple billing accounts for various servers/realms at once in the future, if we need. Otherwise this will just get overwritten during each auth flow.

mateuszmandera replied:
Does `request.session["remote_billing_identities"][remote_realm_uuid]` make sense? Pushed with that naming for now.

timabbott replied:
Yeah let's do that for this PR; we can revise later.

timabbott commented on web/templates/gear_menu_popover.hbs:
```diff
@@ -119,6 +119,14 @@
                     </a>
                 </li>
                 {{/if}}
+                {{#if is_development_environment}}
```
Is it easy to make this `development_environment and (is_owner or billing_administrator)`? That will make it possible to test the intended workflows.

mateuszmandera replied:
Pushing an extra commit for this.

timabbott commented on zerver/lib/exceptions.py:
```diff
@@ -570,3 +571,16 @@ def __init__(self, msg: str, error_type: str) -> None:
+    @staticmethod
+    @override
+    def msg_format() -> str:
+        return _("Missing registration")
```
"Realm not registered" might be most reasonable here.

mateuszmandera replied:
`Organization not registered` perhaps? Otherwise `check-capitalization` complains as usual

timabbott replied:
Sure.

timabbott commented on corporate/views/remote_billing_page.py:
```diff
@@ -0,0 +1,175 @@
+billing_logger = logging.getLogger("corporate.stripe")
+
+
+SIGNED_ACCESS_TOKEN_VALIDITY_IN_SECONDS = 60 * 60 * 24
```
Is there a reason for this to be more than 30 seconds? As best I can tell, the way we use this is such that you're immediately redirected to a page that will use this and store the values in the session, so this is intended to be a very temporary URL construction.

timabbott replied:
I guess in the file upload code path, we have a similar value that's 60s, so maybe we want to use that value but also move to it to a `zproject/default_settings.py` value used in both places; the name `SIGNED_ACCESS_TOKEN_VALIDITY_IN_SECONDS` seems good for it.

`zerver/views/upload.py:        path_id = signer.unsign(signed_data, max_age=timedelta(seconds=60))`

timabbott commented:
The first 3 commits lgtm; still reading the rest.

timabbott commented on corporate/views/remote_billing_page.py:
```diff
@@ -0,0 +1,192 @@
+
+@csrf_exempt
+@typed_endpoint
+def remote_server_billing_entry(
```
Should this be decorated with  `@self_hosting_management_endpoint`?

mateuszmandera replied:
As mentioned above, I didn't put the decorator, because this is a bouncer endpoint - it lives in `zilencer/urls.py` and is queried by the remote server with its bouncer credentials. So this doesn't get queried on the `selfhosting.` subdomain.

I don't think we would want this on  `selfhosting.`  at all, because then server would need to have that subdomain hard-coded in their settings or something. So it seems right to me for this to be requested at `PUSH_NOTIFICATION_BOUNCER_URL`.

timabbott replied:
OK, that makes sense to me. Though maybe we should maybe a parallel `PUSH_BOUNCER_SUBDOMAIN` setting `@push_bouncer_endpoint` decorator? Not sure.

mateuszmandera replied:
Hmm yeah, you mean just adding that as a decorator on all of our bouncer endpoints in general?

Though these have a kinda different subdomain check that happens in the authentication layer in `validate_remote_server`:
```
    if (
        get_subdomain(request) != Realm.SUBDOMAIN_FOR_ROOT_DOMAIN
        and not settings.DEVELOPMENT_DISABLE_PUSH_BOUNCER_DOMAIN_CHECK
    ):
        # Sometimes we may want to test push bouncer logic in development.
        raise JsonableError(_("Invalid subdomain for push notifications bouncer"))
```

Which is a fairly different treatment since we have the bouncer subdomain in `ROOT_SUBDOMAIN_ALIASES` I think? And then we do the check above. So this sounds like it would be a bit of a rework of this current way of doing this.

timabbott replied:
Oh yeah, that's probably fine.

timabbott commented on zerver/decorator.py:
```diff
@@ -74,6 +75,21 @@
+    ) -> HttpResponse:
+        subdomain = get_subdomain(request)
+        if subdomain != settings.SELF_HOSTING_MANAGEMENT_SUBDOMAIN:
+            return render(request, "404.html", status=404)
```
Let's make this `if not settings.DEVELOPMENT or subdomain != settings.SELF_HOSTING_MANAGEMENT_SUBDOMAIN:`;

That way this code is not enabled in production until we feel good enough about its testing to feel comfortable enabling it.

mateuszmandera replied:
Made this tweak. Thought this leaves `remote_server_billing_entry` and `/self-hosted-billing/` endpoints not disabled. Do we want to put `if not settings.DEVELOPMENT` in their code too?

timabbott replied:
Yeah let's do that, just so we don't need to think about the consequences of possible bugs in them.

mateuszmandera replied:
Cool, done

timabbott reviewed (commented):
Reviewed the rest and marked as resolved all the comments that I think are good; there are a few still open. I think at this point, our focus should be on getting CI passing so we can merge this checkpoint; I think everything that's not already looking great here is stuff we can readily do in a follow-up.

andersk commented on zerver/views/push_notifications.py:
```diff
@@ -101,3 +114,45 @@ def send_test_push_notification_api(
+    )
+
+    post_data = {
+        "user": orjson.dumps(dict(user_info)).decode(),
```
Pydantic models should be serialized directly with `user_info.model_dump_json()`, which avoids intermediate untyped `dict` steps.

andersk commented on zerver/views/push_notifications.py:
```diff
@@ -101,3 +114,45 @@ def send_test_push_notification_api(
+    realm_info = get_realms_info_for_push_bouncer(user.realm_id)[0]
+
+    user_info = UserDataForRemoteBilling(
+        uuid=str(user.uuid),
```
Pydantic supports direct serialization of `UUID`; it needn’t be converted to and from `str`.

mateuszmandera replied:
Hmm, fair point here, I made this tweak to `UserDataForRemoteBilling` and (in a prep commit `RealmDataForAnalytics`).

But we still need to convert these to `str()` at the `identity_dict = RemoteBillingIdentityDict(...` step. I tried converting `RemoteBillingIdentityDict` to a `BaseModel`, but `UUID4` just doesn't work with `signing.dumps(dict(...))` (even though it does work with `orjson.dumps`) and throws
```
TypeError: Object of type UUID is not JSON serializable
```

timabbott commented:
Looks like this needs some `nocoverage` comments too.

timabbott commented:
I opened https://github.com/zulip/zulip/pull/27861 to merge the 4 prep commits, since those all looked solid and that should shrink this a bit more.

mateuszmandera commented:
CI passed, so making a final push rebased on top of the now merged https://github.com/zulip/zulip/pull/27861 and with the `[wip]` on the main commit removed.

Aside of the things noted in the discussion above, this also now has some `test_home.py` changes for the new logic in `gear_menu: Add proper logic for when to display remote billing.`

timabbott commented:
Squashed and marked to merge once CI passes. I'll open issues for a few natural follow-ups.
